src/lem/circuit.rs

```diff
@@ -0,0 +1,899 @@
+//! that it can be used with Nova folding to prove evaluations of Lurk expressions.
+//!
+//! Some LEMs may require expensive gadgets, such as Poseidon hashing. So we use
+//! the concept of "slots" to avoid wasting constraints. To explore this idea,
+//! let's use the following LEM as an example:
+//!
+//! ```text
+//! a b c {
+//!     match_tag c {
+//!         Num => {
+//!             let x: Cons = hash2(a, b);
+//!             return (x, x, x);
+//!         },
+//!         Char => {
+//!             let m: Cons = hash2(b, a);
+//!             let n: Cons = hash2(c, a);
+//!             return (m, m, n);
+//!         }
+//!     }
+//! }
+//! ```
+//!
+//! On a first impression, one might think that we need to perform three hashing
+//! operations in the circuit when in fact we can get away with only two. That
+//! is so because interpretation can only follow one of the paths:
+//!
+//! * If it goes through `Num`, we need to get one hash right
+//! * If it goes through `Char`, we need to get two hashes right
+//!
+//! Either way, that's at most two hashes that we really care about. So we say
+//! that we need to allocate two slots. The first slot is for the the hash of `x`
+//! or `m` and the second slot if for the hash of `n`. Let's see a sketch of part
+//! of the circuit:
+//!
+//! ```text
+//!     ┌─────┐        ┌─────┐
+//! s0i0┤slot0├s0  s1i0┤slot1├s1
+//! s0i1┤hash2│    s1i1┤hash2│
+//!     └─────┘        └─────┘
+//! ...
+//! PNum = c.tag == Num
+//! PChar = c.tag == Char
+//!
+//! PNum → a == s0i0
+//! PNum → b == s0i1
+//! PNum → x.hash == s0
+//!
+//! PChar → b == s0i0
+//! PChar → a == s0i1
+//! PChar → m.hash == s0
+//!
+//! PChar → c == s1i0
+//! PChar → a == s1i1
+//! PChar → n.hash == s1
+//! ```
+//!
+//! `PNum` and `PChar` are boolean premises that indicate whether interpretation
+//! went through `Num` or `Char` respectively. They're used as inputs for gadgets
+//! that implement implications (hence the right arrows above). We will talk
+//! about "concrete" vs "virtual" paths later.
+//!
+//! Now we're able to feed the slots with the data that comes from interpretation:
+//!
+//! 1. If it goes through `Num`, we will collect `[[a, b]]` for the preimages of
+//! the slots. So we can feed the preimage of `slot0` with `[a, b]` and the
+//! preimage of `slot1` with anything (we call "dummy values")
+//!
+//! 2. If it goes through `Char`, we will collect `[[b, a], [c, a]]` for the
+//! preimages of the slots. So we can feed the preimage of `slot0` with `[b, a]`
+//! and the preimage of `slot1` with `[c, a]`.
+//!
+//! In the first case, `PNum` will be true, requiring that the conclusions of the
+//! implications for which it is the premise must also be true (which is fine!).
+//! `PChar`, on the other hand, will be false, making the conclusions of the
+//! implications for which it is the premise irrelevant. This is crucial because
+//! we won't have computed `m` nor `n` (for which we will use dummies), thus we
```

Comment:
This is a little ambiguous as written. It read to me as suggesting that dummies will have been used for both `m` and `n`. In fact, I believe `n` will be computed with dummies, but `m` will be computed with `[b, a]` (the `PNum` case). The `m` values are indeed irrelevant because the premise is false, but dummy values will not have been used. For clarity, we should make clear that `m` *is* computed but irrelevant; while `n` uses irrelevant dummies.

Reply:
I've changed the text to better express what I wanted

Reply:
But in your edited text, the parenthetical still seems to suggest that the binding for `m` is a dummy. However, I think that's not what actually happens with the slots optimization. I think there is a real, non-dummy hash in the slot for `m`, but it isn't used.

Reply:
There's indeed a real, non-dummy hash in the *slot* for `m`. But the pointer that's bound to `m` is a dummy one.

Reply:
Really? Okay, I'll try to understand this statement in the context of the implementation. If this means what I think, it seems to imply an unnecessary allocation (since I'm not sure we need 'dummy pointers' at all). However, if that's indeed what is happening we can hopefully optimize away later.

Reply:
Wait, sorry. There's a mistake on what I said. Indeed, there's no dummy pointer at play in the implementation either

Reply:
I've removed that imprecision altogether